### PR TITLE
Update Security Considerations per review feedback

### LINKED ISSUE #241: General Security Questions
I'm a security reviewer from Google Chrome and have some general questions - they almost certainly have good answers and it's very possible I missed finding them in the spec. I also have some specific suggestions which I will open as distinct issues shortly.

1. I think I disagree with the answer given in 2.9 in https://github.com/webmachinelearning/webnn/blob/main/security-privacy.md. This is introducing a new scripting language, built up in js then executed in a number of different interpreters (“cpu”, “gpu” etc.). The complexity of the operations is such that the likelihood of errors that ultimately result in out of bounds accesses is high, and malicious sites will have significant control over operations. This could be addressed more directly in the security/privacy explainer.
2. Operations such as split/slice/squeeze that change the shape of tensors mid-calculation may lead to incorrect assumptions in later operations - for instance if eliding bounds checks this could lead to out of bounds accesses. It would be good for their to be operation level metadata that might be consumed by implementors to help prevent such problems.
3. The universe of operations is likely to vary in future - how will consumers discover which operations are available (short of enumerating them through failures to instantiate)? How will operations be deprecated (for instance if they turn out to be badly implemented?)
4. It feels like .build and .compute should be asynchronous in all cases?
5. New side channels will be made available from shared resources (cpu/gpu). Timeable things should be out of process so incur at least some ipc to achieve anything. Probably not a massive worry when compared with already sharing a cpu between processes running renderers.
6. Verify: Sites must delegate permission to host/run models.
7. Verify: No serialization or caching yet - although this is likely in future.
8. Control over how a model is run - (selecting cpu/gpu/tpu say) - is this too much power for the consuming site - it will for instance make it possible to more directly target a flawed implementation. It's not clear why this is required.

Sub issues:
* #242 
* #243
* #244

### PR DESCRIPTION
**Pull request by anssiko**
Fix #241

***
<a href="https://pr-preview.s3.amazonaws.com/webmachinelearning/webnn/pull/251.html" title="Last updated on Mar 24, 2022, 9:30 AM UTC (a737271)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/webmachinelearning/webnn/251/19a09ab...a737271.html" title="Last updated on Mar 24, 2022, 9:30 AM UTC (a737271)">Diff</a>

### CODE REVIEW

**Comment by anssiko** [comment]
@wchao1115 @huningxin I'm marking this PR as "Ready for review".

My expectation is the first PR that lands will be amended in subsequent PRs as we receive further review feedback. It is fine to have inline issues as reminders in the spec too, but if we're able to fix those via PR review process that'd be of course great.

**Comment by anssiko** [comment]
Here's a direct link to the visual diff of the Security Considerations this PR updates: https://pr-preview.s3.amazonaws.com/webmachinelearning/webnn/251/19a09ab...69db252.html#security

File: index.bs [hunk 1]
```diff
@@ -383,6 +383,41 @@ This API is disabled by default in all cross-origin frames using the [[#permissi
...
+
+Purpose-built Web APIs for measuring high-resolution time mitigate against timing attacks using techniques such as resolution reduction, adding jitter, detection of abuse and API call throttling [[hr-time-3]]. It is expected a typical multi-process architecture implementation of the WebNN API makes use of inter-process communication (IPC) mechanism that inherently adds additional jitter on top of the built-in high-resolution time mitigations. Regardless of these mitigations, implementers are always advised to consider and test their implementations against timing attacks, and in particular if IPC is not used.
+
+## Guidelines for new operations ## {#security-new-ops}
+
+To ensure operations defined in this specification are shaped in a way they can be implemented securely, this section includes guidelines on how operations are expected to be defined to reduce potential for implementation problems. These guidelines are expected to evolve over time to align with industry best practices:
+
+- Avoid reshaping tensors
```
Comment (huningxin) [inline index.bs:406]: I am not quite sure about this one. Do we have any references, say ONNX document?

File: index.bs [hunk 2]
```diff
@@ -383,6 +383,41 @@ This API is disabled by default in all cross-origin frames using the [[#permissi
 
 This API allows creation of an {{MLContext}} from a {{GPUDevice}} or {{WebGLRenderingContext}} defined by WebGPU and WebGL specifications respectively. See <a href="https://gpuweb.github.io/gpuweb/#security">WebGPU Security Considerations</a> and <a href="https://www.khronos.org/registry/webgl/specs/latest/1.0/#4">WebGL Security Consideration</a> for more information regarding security characteristics of these contexts.
 
+Some operations such as {{split()}}, {{slice()}} and {{squeeze()}} that change the shape of tensors mid-calculation may lead to incorrect assumptions in later operations and possibly lead to out-of-bounds access attempts. Implementers are adviced to anticipate such out-of-bounds access issues and add appropriate mitigations.
```
Comment (huningxin) [inline index.bs:386]: As @wchao1115 [mentioned](https://github.com/webmachinelearning/webnn/issues/243) in #243, the shape inference is done during the graph building stage. When compute the graph, the input shape of each op is finalized. Probably we could make it clear about that.

**Review by huningxin — COMMENTED** [review]
Thanks for making this PR, @anssiko !

File: index.bs (hunk 1 above)
Comment (anssiko) [reply index.bs:406]:
```suggestion
```

The [ONNX document](https://lists.w3.org/Archives/Public/www-archive/2021Nov/att-0000/W3C_Adding_new_Operators.pdf) does not mention this. I believe this must be my misinterpretation/oversimplification of [our discussion](https://www.w3.org/2022/01/27-webmachinelearning-minutes.html#t03).

I'm suggesting we drop this bullet unless we can reword this to make sense in this context.

File: index.bs (hunk 2 above)
Comment (anssiko) [reply index.bs:386]:
```suggestion
Some operations such as {{split()}}, {{slice()}} and {{squeeze()}} that change the shape of tensors mid-calculation may lead to incorrect assumptions in later operations and possibly lead to out-of-bounds access attempts. Implementers are advised to anticipate such out-of-bounds access issues and add appropriate mitigations.

This concern is in part mitigated by means of API design. Once the graph is fully constructed and compiled, the input shapes into each of the operations in the graph are inferred and finalized. The bounds checking occurs when the compute method is invoked that executes the graph against the actual data. No actual data is bound to the compiled graph before this stage. It is the implementation's responsibility to make sure proper bounds checking occurs against the shapes of the data already inferred by that time.
```
I added and adapted @wchao1115's explanation (second paragraph). Please review and suggest changes.

**Comment by anssiko** [comment]
@dontcallmedom @wchao1115 PTAL prior to the WG's call this week.

File: index.bs [hunk 3]
```diff
@@ -383,6 +383,41 @@ This API is disabled by default in all cross-origin frames using the [[#permissi
...
+
+Issue: Document operations susceptible to out-of-bounds access as a guidance to implementers.
+
+As a future-proofing, the API design allows certain operation (annotated with "can be generically emulated") defined in this specification to be deprecated for security, performance, or other reasons without breaking compability. This is made possible by high-level functions that are defined in terms of smaller primitive operations defined in this specifications. This enables a native implementation of a high-level function to be replaced with a polyfill implementation.
+
+Issue: Investigate side channel attack feasibility considering the current state where CPU is shared between processes running renderers.
+
+In order to not allow an attacker to target a specific implementation that may contain a flaw, the [[#programming-model-device-selection]] mechanism is a hint only, and the concrete device selection is left to the implementation. As a further mitigation, no device enumeration mechanism is defined.
```
Comment (dontcallmedom) [inline index.bs:394]:
```suggestion
In order to not allow an attacker to target a specific implementation that may contain a flaw, the [[#programming-model-device-selection]] mechanism is a hint only, and the concrete device selection is left to the implementation - a user agent could for instance choose never to run a model on a device with known vulnerabilities. As a further mitigation, no device enumeration mechanism is defined.
```

File: index.bs [hunk 4]
```diff
@@ -383,6 +383,41 @@ This API is disabled by default in all cross-origin frames using the [[#permissi
...
+
+Issue: Hinting partially mitigates the concern. Investigate additional mitigations.
+
+The API design minimizes the attack surface for the compiled computational graph. The {{MLGraphBuilder}} interface that hosts the various operations is a data definition API and as such doesn't execute anything, only constructs data. What follows, is that the potential for an attack is limited to when binding the data to the graph before executing it by invoking the {{MLGraph/compute()}} method. This enables implementers to focus on hardening the {{MLGraph/compute()}} method. For example, by making sure it honors the boundary of data and fails appropriately when the bounds are not respected.
```
Comment (dontcallmedom) [inline index.bs:398]: this will need an editorial update if/when #257 lands

File: index.bs [hunk 5]
```diff
@@ -383,6 +383,41 @@ This API is disabled by default in all cross-origin frames using the [[#permissi
...
+
+Purpose-built Web APIs for measuring high-resolution time mitigate against timing attacks using techniques such as resolution reduction, adding jitter, detection of abuse and API call throttling [[hr-time-3]]. It is expected a typical multi-process architecture implementation of the WebNN API makes use of inter-process communication (IPC) mechanism that inherently adds additional jitter on top of the built-in high-resolution time mitigations. Regardless of these mitigations, implementers are always advised to consider and test their implementations against timing attacks, and in particular if IPC is not used.
```
Comment (dontcallmedom) [inline index.bs:400]:
```suggestion
Purpose-built Web APIs for measuring high-resolution time mitigate against timing attacks using techniques such as resolution reduction, adding jitter, detection of abuse and API call throttling [[hr-time-3]]. The practical deployment of WebNN implementations are likely to bring enough jitter to make timing attacks impractical (e.g. because they would use IPC) but implementers are advised to consider and test their implementations against timing attacks.
```

File: index.bs [hunk 6]
```diff
@@ -383,6 +383,41 @@ This API is disabled by default in all cross-origin frames using the [[#permissi
...
 
+Some operations such as {{split()}}, {{slice()}} and {{squeeze()}} that change the shape of tensors mid-calculation may lead to incorrect assumptions in later operations and possibly lead to out-of-bounds access attempts. Implementers are adviced to anticipate such out-of-bounds access issues and add appropriate mitigations.
+
+Issue: Document operations susceptible to out-of-bounds access as a guidance to implementers.
+
+As a future-proofing, the API design allows certain operation (annotated with "can be generically emulated") defined in this specification to be deprecated for security, performance, or other reasons without breaking compability. This is made possible by high-level functions that are defined in terms of smaller primitive operations defined in this specifications. This enables a native implementation of a high-level function to be replaced with a polyfill implementation.
```
Comment (wchao1115) [inline index.bs:390]: I'm not sure I agree with this statement. How do emulated operations pose security threat and should be deprecated in the future if it's an implementation choice of the API?

File: index.bs (hunk 3 above)
Comment (wchao1115) [inline index.bs:394]: It's actually important for the applications to understand how a device is selected and that the device selection rule, if we  would have one, should be a normative behavior and not merely informative.

File: index.bs (hunk 1 above)
Comment (wchao1115) [reply index.bs:406]: Reshaping is quite common in an ML graph. I'm not entirely sure why it should be discouraged from a security standpoint.

File: index.bs (hunk 2 above)
Comment (anssiko) [reply index.bs:386]: Suggested change updated accordingly.

File: index.bs (hunk 3 above)
Comment (anssiko) [reply index.bs:394]: This refers to the following ops that can be generically emulated (and for which we have informative "can be generically emulated" text in place):

`batchNormalization, clamp, elu, gemm, gru, gruCell, hardSigmoid, hardSwish, instanceNormalization, leakyRelu, linear, relu, sigmoid, softmax, softplus, softsign, split, tanh`

Suggested change updated. I did not enumerate these ops, but refer to them as "certain operations that can be generically emulated".

File: index.bs (hunk 1 above)
Comment (anssiko) [reply index.bs:406]: It sounds like we'll simply drop this bullet as suggested.

**Comment by anssiko** [comment]
@wchao1115 @dontcallmedom @huningxin thanks for your comments, we're iterating toward a mergeable PR.

For discussion on this week's call.

File: index.bs (hunk 6 above)
Comment (anssiko) [reply index.bs:390]:
```suggestion
As a future-proofing measure, the API design allows certain operations that can be generically emulated to be deprecated for security, performance, or other reasons without breaking compatibility. This is made possible by high-level functions that are defined in terms of smaller primitive operations defined in this specifications. This enables a native implementation of a high-level function to be replaced with a polyfill implementation.
```

Comment (anssiko) [reply index.bs:390]: (Reposting to the right line 390:)

This refers to the following ops that can be generically emulated (and for which we have informative "can be generically emulated" text in place):

`batchNormalization, clamp, elu, gemm, gru, gruCell, hardSigmoid, hardSwish, instanceNormalization, leakyRelu, linear, relu, sigmoid, softmax, softplus, softsign, split, tanh`

Suggested change updated. I did not enumerate these ops, but refer to them as "certain operations that can be generically emulated".

**Comment by anssiko** [comment]
I incorporated all the suggestions made in this PR to trigger a PR preview update.

See: [Preview](https://pr-preview.s3.amazonaws.com/webmachinelearning/webnn/pull/251.html) | [Diff](https://pr-preview.s3.amazonaws.com/webmachinelearning/webnn/251/19a09ab...a737271.html)

**Review by huningxin — APPROVED** [review]
